compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1506,6 +1507,29 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                         if !fcx.tcx.features().unsized_locals {
                             unsized_return = self.is_return_ty_unsized(fcx, blk_id);
                         }
+                        if let Some(expression) = expression {
+                            match expression.kind {
+                                hir::ExprKind::Loop(
+                                    loop_blk,
+                                    _loop_opt_label,
+                                    _loop_k,
+                                    _loop_span,
+                                ) => {
```

Comment:
We can use let-chains to remove another level of nesting:

```suggestion
                        if let Some(expression) = expression
                            && let hir::ExprKind::Loop(loop_blk,..) = expression.kind
                        {

```